src/analysis/processing/qgsalgorithmspatialjoin.cpp

```diff
@@ -0,0 +1,375 @@
+{
+
+  std::unique_ptr<QgsProcessingFeatureSource> mBaseSource( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !mBaseSource )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+
+  std::unique_ptr<QgsProcessingFeatureSource> mJoinSource( parameterAsSource( parameters, QStringLiteral( "JOIN" ), context ) );
+  if ( !mJoinSource )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "JOIN" ) ) );
+
+  QString prefix = parameterAsString( parameters, QStringLiteral( "PREFIX" ), context );
+
+  std::unique_ptr< QgsFeatureSink > joinedFeatures( parameterAsSink( parameters, QStringLiteral( "JOINED" ), context, joinedFeaturesId, mBaseSource->fields(),
+      mBaseSource->wkbType(), mBaseSource->sourceCrs(), QgsFeatureSink::RegeneratePrimaryKey ) );
+
+  bool mDiscardNonMatching = parameterAsBoolean( parameters, QStringLiteral( "DISCARD_NONMATCHING" ), context );
+
+  if ( mDiscardNonMatching )
+    std::unique_ptr< QgsFeatureSink > unjoinedFeatures( parameterAsSink( parameters, QStringLiteral( "UNJOINED" ), context, unjoinedFeaturesId, mBaseSource->fields(),
+        mBaseSource->wkbType(), mBaseSource->sourceCrs(), QgsFeatureSink::RegeneratePrimaryKey ) );
+
+  mJoinMethod = parameterAsEnum( parameters, QStringLiteral( "METHOD" ), context );
+
+  QList<int> mPredicates = parameterAsEnums( parameters, QStringLiteral( "PREDICATE" ), context );
+
+  QgsFields outFields2;
+  QgsAttributeList fields2Indices;
+  if ( mFieldNames.empty() )
+  {
+    outFields2 = mJoinSource->fields();
+    fields2Indices.reserve( outFields2.count() );
+    for ( int i = 0; i < outFields2.count(); ++i )
+    {
+      fields2Indices << i;
+    }
+  }
+  else
+  {
+    fields2Indices.reserve( mFieldNames.count() );
+    for ( const QString &field : mFieldNames )
+    {
+      int index = mJoinSource->fields().lookupField( field );
+      if ( index >= 0 )
+      {
+        fields2Indices << index;
+        outFields2.append( mJoinSource->fields().at( index ) );
+      }
+    }
+  }
+
+  if ( !prefix.isEmpty() )
+  {
+    for ( int i = 0; i < outFields2.count(); ++i )
+    {
+      outFields2[ i ].setName( prefix + outFields2[ i ].name() );
+    }
+  }
+  for ( int i = 0; i < fields2Indices.count() + 6; ++i )
+    emptyAttrs << QVariant();
+
+
+  QgsAttributeList fields2Fetch = fields2Indices;
+  fields2Fetch << fields2Indices;
+
+  static const QgsFields mOutFields = QgsProcessingUtils::combineFields( mBaseSource->fields(), outFields2 );
+
+  return true;
+}
+
+
+QVariantMap QgsSpatialJoinAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+
+  // make spatial index
+  QgsFeatureIterator joinIter = mJoinSource->getFeatures( QgsFeatureRequest().setDestinationCrs( mBaseSource->sourceCrs(), context.transformContext() ).setSubsetOfAttributes( fields2Fetch ) );
```

Comment:
Missing many parts from the original python algorithm -- please restore/port these too

Reply:
Were you referring to the counting function? That should be in now.